src/gui/SampleWaveform.cpp

```diff
@@ -28,64 +28,39 @@ namespace lmms::gui {
 
 void SampleWaveform::visualize(Parameters parameters, QPainter& painter, const QRect& rect)
 {
-	const int x = rect.x();
-	const int height = rect.height();
-	const int width = rect.width();
-	const int centerY = rect.center().y();
+	if (!parameters.buffer || parameters.size == 0) { return; }
 
-	const int halfHeight = height / 2;
+	const auto centerY = rect.center().y();
+	const auto halfHeight = rect.height() / 2;
 
 	const auto color = painter.pen().color();
 	const auto rmsColor = color.lighter(123);
 
-	const float framesPerPixel = std::max(1.0f, static_cast<float>(parameters.size) / width);
+	const auto samplesPerPixel = std::max(1.0f, static_cast<float>(parameters.size) / rect.width());
+	const auto numPixelsToDraw = std::min(static_cast<int>(parameters.size), rect.width());
 
-	constexpr float maxFramesPerPixel = 512.0f;
-	const float resolution = std::max(1.0f, framesPerPixel / maxFramesPerPixel);
-	const float framesPerResolution = framesPerPixel / resolution;
+	const auto compFn = [](const SampleFrame& a, const SampleFrame& b) { return a.average() < b.average(); };
+	const auto squaredSumFn = [](const float acc, const SampleFrame& x) { return acc + x.average() * x.average(); };
 
-	size_t numPixels = std::min(parameters.size, static_cast<size_t>(width));
-	auto min = std::vector<float>(numPixels, 1);
-	auto max = std::vector<float>(numPixels, -1);
-	auto squared = std::vector<float>(numPixels, 0);
-
-	const size_t maxFrames = static_cast<size_t>(numPixels * framesPerPixel);
-
-	auto pixelIndex = std::size_t{0};
-
-	for (auto i = std::size_t{0}; i < maxFrames; i += static_cast<std::size_t>(resolution))
+	for (auto i = 0; i < numPixelsToDraw; i++)
 	{
-		pixelIndex = i / framesPerPixel;
-		const auto frameIndex = !parameters.reversed ? i : maxFrames - i;
+		const auto start = parameters.buffer + static_cast<int>(std::floor(i * samplesPerPixel));
+		const auto end = parameters.buffer + static_cast<int>(std::ceil((i + 1) * samplesPerPixel));
```

Comment:
Let's say `parameters.size` = 52532 and `rect.width()` = 730. If this is the case, then `numPixelsToDraw` = 730, `samplesPerPixel` = 71.961643836f. We will enter a loop where `i` can be anywhere from 0 to 729 + 1. If we multiply  `(729 + 1)×71,961643836 = 52532.00000028` This will be rounded up to `52533` which is wrong. The max should be `52531`, because that is the last index inside the buffer.

What happens if `parameters.size` = 730 and `rect.width()` = 52532: `numPixelsToDraw` = 730, `samplesPerPixel` = 1.0f. `end = (729 + 1) * 1.0f = 730`. The max should be `729`, because that is the last index of the buffer.

Reply:
This if should fix the issue while still looping through the samples.
```suggestion
		const SampleFrame* start = parameters.buffer + static_cast<int>(std::floor(i * samplesPerPixel));
		const auto endIndex = static_cast<int>(std::ceil((i + 1) * samplesPerPixel));
		endIndex = endIndex > parameters.size - 1 ? parameters.size - 1 : endIndex;
		const SampleFrame* end = parameters.buffer + endIndex;
		
```

Reply:
> If we multiply (729 + 1)×71,961643836 = 52532.00000028 This will be rounded up to 52533 which is wrong. The max should be 52531, because that is the last index inside the buffer.

> What happens if parameters.size = 730 and rect.width() = 52532: numPixelsToDraw = 730, samplesPerPixel = 1.0f. end = (729 + 1) * 1.0f = 730. The max should be 729, because that is the last index of the buffer.

`end` is *supposed* to point one after the last index inside the buffer. Your solution unfortunately would instead ignore the last index inside the buffer, causing a bug. `52533` is correct because it points to one after the last index inside the buffer.

Referring to [cppreference](https://en.cppreference.com/w/cpp/algorithm/accumulate), it states that `std::accumulate` works in the range `[first, last)`. This range excludes `last`. Half-open intervals are usually what C++ algorithms are based upon that work across a range. `std::minmax_element` works the [same way](https://en.cppreference.com/w/cpp/algorithm/minmax_element).

Reply:
Hm, however, `52533` is infact wrong actually. It should've been `52532`.

Reply:
Seems like giving `samplesPerPixel` more accuracy by turning it into a `double` fixes the issue.